bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -461,87 +489,343 @@ public Boolean isEnabled(String ruleUID) {
+            }
+        } else {
+            throw new IllegalArgumentException(
+                    String.format("The rule '%s' is not updated, not matching with any existing rule", uid));
+        }
+    }
+
+    @Override
+    protected void onAddElement(Rule element) throws IllegalArgumentException {
+        try {
+            validateModuleIDs(element, element.getModules(Module.class));
+            resolveConfigurations(element);
+        } catch (Exception e) {
+            logger.debug("Added rule '{}' is invalid", element.getUID(), e);
+        }
+    }
+
+    @Override
+    protected void onUpdateElement(Rule oldElement, Rule element) throws IllegalArgumentException {
+        try {
+            validateModuleIDs(element, element.getModules(Module.class));
+            resolveConfigurations(element);
+        } catch (Exception e) {
+            logger.debug("The new version of updated rule '{}' is invalid", element.getUID(), e);
+        }
+    }
+
+    /**
+     * This method is used to add or update the {@link Rule} into {@link RuleEngine}.
+     *
+     * @param rulethe {@link Rule} to be added/updated
+     * @param statusInfo the initial {@link RuleStatusInfo} for the specified rule, determined by the
+     *            {@link RuleRegistry}.
+     */
+    private void addToRuleEngine(Rule rule, RuleStatusInfo statusInfo) {
+        RuleEngine ruleEngine = null;
+        synchronized (this) {
+            ruleEngine = this.ruleEngine;
+            ruleStatuses.put(rule.getUID(), statusInfo);
+        }
+        if (ruleEngine != null) {
+            ruleEngine.addRule(rule, statusInfo);
+        }
+    }
+
+    /**
+     * This method serves to resolve and normalize the {@link Rule}s configuration values and its module configurations.
+     *
+     * @param rule the {@link Rule}, whose configuration values and module configuration values should be resolved and
+     *            normalized.
+     */
+    private void resolveConfigurations(Rule rule) {
+        List<ConfigDescriptionParameter> configDescriptions = rule.getConfigurationDescriptions();
+        Configuration configuration = rule.getConfiguration();
+        ConfigurationNormalizer.normalizeConfiguration(configuration,
+                ConfigurationNormalizer.getConfigDescriptionMap(configDescriptions));
+        Map<String, Object> configurationProperties = configuration.getProperties();
+        if (rule.getTemplateUID() == null) {
+            String uid = rule.getUID();
+            try {
+                validateConfiguration(configDescriptions, new HashMap<>(configurationProperties));
+                resolveModuleConfigReferences(rule.getModules(Module.class), configurationProperties);
+                ConfigurationNormalizer.normalizeModuleConfigurations(rule.getModules(Module.class),
+                        moduleTypeRegistry);
+            } catch (Exception e) {
```

Comment:
We should not catch generic Exceptions. Can you make this more specific?

Reply:
OK